Sprint_16

Wilwous opened the pull request:
Создал функционал создания и добавления в трекер категорий с сохранением в Core Data по системе проектирования MVVM
Добавлен онбординг
Создал кастомный TextField для переиспользования

Поправил обновление методов обработки завершения и отмены трекеров
fПоправил создание нерегулярного события только на текущий день

Commit:
Добавил ассеты цветов и лого.
Сверстал лаунчскрин.

Commit:
feat: Добавлен TabBarViewController

- Добавлены вкладки "Трекеры" и "Статистика" в TabBarViewController
- Добавлена разделяющая линия

feat: Добавлен TrackerViewController

- Реализован поиск в TrackerViewController
- Добавлена возможность выбора даты
- Добавлена кнопка для добавления трекеров
- Добавлены лейблы для различных элементов в TrackerViewController

feat: Добавлен StatisticsViewController

chore: Добавлена иконка приложения и прочие ассеты.

Commit:
fix: Исправление и доработка форматирования даты

- Добавлены констрейны ширины для корректного отображения форматирования даты в цифрах
- Удалена дублирующая строка кода

Commit:
feat: Добавлены структуры Tracker, TrackerCategory и TrackerRecord

- Создана сущность Tracker для хранения информации о трекере с уникальным идентификатором, названием, цветом, эмоджи и расписанием.
- Введена TrackerCategory, представляющая категорию трекеров с заголовком и массивом трекеров внутри.
- Добавлена TrackerRecord для хранения записей о выполнении трекера на определенную дату с id трекера и датой выполнения.

feat: Обновлены свойства и методы в TrackersViewController

- Добавлены свойства categories и completedTrackers для хранения списка категорий и выполненных трекеров соответственно.
- Реализован механизм добавления новых трекеров в категории и отметок о выполнении в completedTrackers.

Commit:
fix

Commit:
feat: Добавлена функциональность кнопки добавления трекера

- Добавлена приватная функция addTrackerButtonTapped для обработки нажатия на кнопку добавления трекера.
- Реализован механизм перехода к экрану создания трекера при нажатии на кнопку.

Добавлен экран для создания трекеров

- Создан экран TrackerCreator для создания новых трекеров.
- Добавлены элементы интерфейса: метка "Создание трекера", кнопки "Привычка" и "Нерегулярное событие".
- Реализованы констрейнты для размещения элементов интерфейса на экране.

chore: Переименована переменная для лучшей читаемости

Commit:
UPD:
feat: Добавлена функциональность кнопки добавления трекера
- Добавлена приватная функция addTrackerButtonTapped для обработки нажатия на кнопку добавления трекера.
- Реализован механизм перехода к экрану создания трекера при нажатии на кнопку.

Добавлен экран для создания трекеров

- Создан экран TrackerCreator для создания новых трекеров.
- Добавлены элементы интерфейса: метка "Создание трекера", кнопки "Привычка" и "Нерегулярное событие".
- Реализованы констрейнты для размещения элементов интерфейса на экране.

chore: Переименована переменная для лучшей читаемости

Commit:
feat: Добавлен экран создания привычек

- Создан экран HabitCreation для создания новых привычек.
- Добавлены элементы интерфейса: метка "Новая привычка", текстовое поле для ввода названия привычки, сообщение об ограничении символов, таблица для выбора категории и расписания, кнопки "Отменить" и "Создать".
- Реализованы констрейнты и настройка интерфейса для корректного отображения элементов на экране.
- Реализована таблица с ячейками для выбора категории и расписания привычки.

chore: Переорганизация структуры проекта

Commit:
feat: Добавлен экран создания расписания

- Создан класс TimetableCreation для отображения экрана создания расписания.
- Реализован выбор дней недели с помощью UITableView.
- Добавлена кнопка "Готово" для завершения выбора дней недели.
- Добавлен протокол TimetableCreationDelegate для обратной связи с родительским контроллером.
- Создан класс TimetableTableView для ячейки таблицы с днями недели.
- Добавлена возможность выбора дней недели с помощью UISwitch в ячейке таблицы.

Commit:
feat: Добавлена коллекция трекера.
- Реализовано отображение заглушки.
- Добавлена возможность отметки трекера для пометки выполненных задач.
- Реализовано всплывающее сообщение о ограничение на количество символов.
- Добавлена функциональность фильтрации по дате и поиску для удобства работы с данными.

refactor:
- Переименованы переменные для более ясного понимания их назначения.
- Добавлены комментарии Mark для улучшения навигации и структурирования кода.
- Организован код и файлы для улучшения читаемости и поддерживаемости проекта.
- Пересмотрены и скорректированы констрейнты для улучшения внешнего вида интерфейса.

Commit:
feat: Добавлена функциональность создания и отмены трекера.
Введен алерт для информирования пользователя об ошибке при отсутствии названия трекера.
Реализована смена состояния кнопки при готовности создания трекера.

Commit:
fix: Поправил смену состояние кнопки создать.

Commit:
fix: - Устранена проблема с нажатием по таблице в выборе расписания трекера
- Исправлена проблема с ассетами: теперь появляется галочка завершения трекера и заглушка при отсутствии трекера
- Исправлена проблема, при которой страница создания трекера открывалась сразу
- Улучшена фильтрация поиска по выбранному дню
- Исправлено расположение кнопок создания и отмены трекера
- Улучшен функционал создания нового трекера

refactor: - Заменен поиск трекера с использованием UISearchController на UISearchBar
- Проведена реструктуризация кода для сокращения его объема

Commit:
fix: - Поправил ассеты
- Поправил нажатие на таблицу

Commit:
refactor: Убрал лишний код

Commit:
feature: - Создан экран нерегулярного события
- Добавлена коллекция цветов и эмодзи для создания трекера

refactor:
- Произведен рефакторинг кода для переиспользования

fix:
- Исправлен цвет лейбла трекера в соответствии с макетом

Commit:
feature: - Подключен Core Data для управления данными трекеров
- Добавлены классы для работы с Core Data: CoreDataStack, TrackerStore, TrackerCategoryStore, TrackerRecordStore
- Реализована поддержка хранения и управления трекерами, категориями и записями

Commit:
feature: - Добавлено создание нерегулярного события с сохранением в CoreData

refactor:
- Убран скролл эмодзи и цветов в создании трекера
- Сделан fetchedResultsController опциональным
- Проведена корректировка кода

fix:
- Исправлен цвет лейбла трекера в соответствии с макетом

Commit:
feature: - Добавлен Onboarding

Commit:
feat: Создал функционал создания и добавления в трекер категорий с сохранением в Core Data по системе проектирования MVVM
feat: Создал кастомный TextField для переиспользования
fix: Поправил обновление методов обработки завершения и отмены трекеров
fix: Исправил отображение надписи в онбординг
fix: Поправил создание нерегулярного события только на текущий день

Helvetii951 commented on Tracker/CoreDateService/CoreDataStack.swift:
```diff
@@ -0,0 +1,41 @@
+import Foundation
+import CoreData
+
+final class CoreDataStack {
+    
+    // MARK: - Core Data stack
+    static let shared = CoreDataStack()
+    
+    lazy var persistentContainer: NSPersistentContainer = {
+        let container = NSPersistentContainer(name: "Model")
+        let description = container.persistentStoreDescriptions.first
+               description?.shouldMigrateStoreAutomatically = true
+               description?.shouldInferMappingModelAutomatically = true
+        container.loadPersistentStores(completionHandler: { (storeDescription, error) in
+            if let error = error as NSError? {
+                fatalError("Unresolved error \(error), \(error.userInfo)")
+            }
+        })
+        return container
+    }()
```
Здесь стоит добавить приватный инициализатор. Кто-то может создать еще один `CoreDataStack()` и получить новый экземпляр `NSPersistentContainer`.

Часто при использовании `static let shared` пишут приватный инициализатор. Тем самым реализуя паттерн [singleton](https://theswiftdev.com/swift-singleton-design-pattern/).

Helvetii951 commented on Tracker/CoreDateService/CoreDataStack.swift:
```diff
@@ -0,0 +1,41 @@
+        let description = container.persistentStoreDescriptions.first
+               description?.shouldMigrateStoreAutomatically = true
+               description?.shouldInferMappingModelAutomatically = true
+        container.loadPersistentStores(completionHandler: { (storeDescription, error) in
+            if let error = error as NSError? {
+                fatalError("Unresolved error \(error), \(error.userInfo)")
+            }
+        })
```
`fatalError` лучше никогда не использовать. Приложение в продакшене просто упадет. В таких случаях пишут ошибки в логи или сервисы статистики и событий, а затем быстро решают.

Helvetii951 commented on Tracker/TrackerManager/CategoryListViewController/AddCategoryViewController.swift:
```diff
@@ -0,0 +1,128 @@
+final class AddCategoryViewController: UIViewController {
+    
+    // MARK: - Properties
+    var onCategoryAdded: ((String) -> Void)?
+    private var viewModel = AddCategoryViewModel()
+    
```
Для гибкости и тестируемости кода желательно объявлять вью-модель протоколом и как-нибудь инжектить.

[Неплохая статья по внедрению зависимостей](https://habr.com/ru/companies/tinkoff/articles/546360/).

Helvetii951 commented on Tracker/TrackerManager/Onboarding/OnboardingPage.swift:
```diff
@@ -0,0 +1,31 @@
+
+import Foundation
+
+enum OnboardingPage: Int, CaseIterable {
+    case pageOne = 0
+    case pageTwo
+
```
Хорошее решение обернуть инфу о странице в енум👍

Helvetii951 commented on Tracker/TrackerManager/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,146 @@
+    }
+    
+    // MARK: - Navigation
+    private func switchToMainInterface() {
+        if let window = view?.window {
+            let mainTabBarController = TabBarViewController()
+            window.rootViewController = mainTabBarController
+            UIView.transition(
+                with: window,
+                duration: 0.5,
+                options: .transitionCrossDissolve,
+                animations: nil,
+                completion: nil
+            )
+        }
+    }
```
Лучше забрать у контроллера онбординга обязанность навигации и отдать. ее тому, кто создал контроллер(в нашем случае это `SceneDelegate`). Передаем событие нажатия кнопки через, например, замыкание, а `SceneDelegate` уже пусть сам решает что делать.

Helvetii951 commented on Tracker/TrackerManager/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,146 @@
+
+final class OnboardingViewController: UIPageViewController {
+    
+    lazy var pageControl: UIPageControl = {
+        let page = UIPageControl()
+        page.currentPage = 0
+        page.currentPageIndicatorTintColor = .ypGray
+        page.currentPageIndicatorTintColor = .ypBlackDay
+        page.numberOfPages = OnboardingPage.allCases.count
+        
+        return page
+    }()
+    
+    lazy var continueButton: UIButton = {
+        let button = UIButton(type: .system)
+        button.setTitle("Вот это технологии!", for: .normal)
+        button.titleLabel?.font = .systemFont(ofSize: 16, weight: .medium)
+        button.setTitleColor(.white, for: .normal)
+        button.backgroundColor = .black
+        button.layer.cornerRadius = 16
+        button.addTarget(
+            self,
+            action: #selector(continueButtonTapped),
+            for: .touchUpInside
+        )
+        
+        return button
+    }()
```
Свойства можно сделать приватными. Рекомендую всегда писать функции и свойства с минимальной областью видимости, а затем менять ее, если нужно.

Helvetii951 commented on Tracker/AppDelegate.swift:
```diff
@@ -10,10 +10,16 @@ import UIKit
 
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
-        // Override point for customization after application launch.
+
+        window = UIWindow(frame: UIScreen.main.bounds)
+        
+        let tabBarController = TabBarViewController()
+        window?.rootViewController = tabBarController
+        window?.makeKeyAndVisible()
+
```
Это очень нужно убрать. Установка стартового экрана уже проводится в `SceneDelegate`.

Helvetii951 commented on Tracker/Helpers/ViewState.swift:
```diff
@@ -0,0 +1,13 @@
+
+import Foundation
+
+enum ViewState {
+    case empty
+    case populated
+}
```
Тоже хорошее решение. Но в нэйминге стоит уточнить какая именно вью. В проекте их может быть очень много.

Commit:
refactor: CategoryViewController и AddCategoryViewModel

- Интегрирована CategoryViewModel для управления состоянием категорий через биндинги.
- Логика управления категориями перенесена из CategoryViewController в вьюмодель.
- Добавлен биндинг обновлений вьюмодели с UI.
- Улучшено отображение пустого списка категорий.
- Создана AddCategoryViewModel для добавления новых категорий и передачи данных в CategoryViewModel.

refactor: OnboardingViewController и SceneDelegate

- Удалена ответственность за навигацию из OnboardingViewController.
- Добавлено замыкание onContinue для передачи события завершения онбординга.
- Обновлен метод continueButtonTapped для установки флага завершения онбординга и вызова замыкания onContinue.
- В SceneDelegate добавлена проверка флага завершения онбординга при запуске приложения.
- Реализовано переключение между онбордингом и основной частью приложения на основе значения флага.
- Метод switchToMainInterface перенесен в SceneDelegate для управления переходом к основной части приложения.

fix: Устранение ненужной установки стартового экрана в AppDelegate

- Удалена избыточная логика установки стартового экрана в AppDelegate.

feat: Добавление приватного инициализатора для CoreDataStack

- Добавлен приватный инициализатор для CoreDataStack, чтобы предотвратить создание экземпляров вне класса и обеспечить использование синглтона.

refactor: Обновления кода

- Улучшено оформление и структура кода.

Wilwous commented:
@Helvetii951                Произведены изменения с учетом замечаний и рекомендаций.

CategoryViewController и AddCategoryViewModel:
- Интеграция CategoryViewModel для управления категориями через биндинги.
- Перенос логики управления категориями в вьюмодель.
- Улучшение отображения пустого списка категорий.
- Создание AddCategoryViewModel для добавления категорий.

OnboardingViewController и SceneDelegate:
- Удаление ответственности за навигацию из OnboardingViewController.
- Добавление замыкания onContinue для завершения онбординга.
- Проверка флага завершения онбординга в SceneDelegate и управление переходом.

Удаление ненужной установки стартового экрана в AppDelegate
Приватный инициализатор для CoreDataStack
Улучшение структуры кода и комментариев.
